mypy/test/data.py

```diff
@@ -470,14 +471,29 @@ def pytest_pycollect_makeitem(collector: Any, name: str, obj: Any) -> Any:
     return MypyDataSuite(name, parent=collector)
 
 
+# An instance of this class is created for each subclass of DataSuite encountered by pytest;
+# pytest.Class machinery sets instance attribute obj: Type[DataSuite] to that subclass.
+# Beyond this, we should not rely on obj-related machinery, since it's undocumented and implicit.
 class MypyDataSuite(pytest.Class):  # type: ignore  # inheriting from Any
+    # Will yield collectors that represent individual case files.
+    def collect(self) -> Iterator['MypyDataCaseFile']:
+        for case_file in self.obj.case_files:
+            yield MypyDataCaseFile(case_file, self)
+
+
+# An instance of this class is tied (via .name and .parent) to a specific case file.
+class MypyDataCaseFile(pytest.Collector):  # type: ignore  # inheriting from Any
     def collect(self) -> Iterator['MypyDataCase']:
-        for case in self.obj.cases():
+        for case in parse_test_cases(os.path.join(test_data_prefix, self.name),
+                                     None, test_temp_dir, True):
             yield MypyDataCase(case.name, self, case)
 
 
+# An instance of this class is tied (via .obj) to a specific case.
+# Note that despite the use of the obj attribute, the obj-related pytest machinery is not invoked.
+# TODO: rename obj into case.
```

Comment:
Also before we merge we should remember to resolve any TODO items like this one :)

I'm not actually sure what you mean by "the obj-related pytest machinery", though -- possibly because I've swapped out some of how that API works from my head. If this still applies after the rework to file-based collection, can you say a bit more about it?